### DISCUSSION TITLE
OutOfMemoryException when opening a pdf file with a 4k monitor

### DISCUSSION THREAD

**Discussion by abid76** [cefsharp/CefSharp#4275] [General]
- **What version of the product are you using?**
    105.3.390

- **What architecture x86 or x64?**
    x86
	
- **What version of .Net?**
    .Net 4.8
    
- **On what operating system?**
    Win11

- **Are you using `WinForms`, `WPF` or `OffScreen`?**
    WPF
    
- **What steps will reproduce the problem?**
    Wasn't able to reproduce this error.

- **What is the expected output? What do you see instead?**
    Customer opens a **pdf file**. Instead of displaying the file, an `OutOfMemoryException` raises 

Unfortunately I could not reproduce this error, but it occures quite regularly on a customers's machine where they are using a 4k monitor.

Stacktrace:
> System.OutOfMemoryException: Nicht genügend Speicher verfügbar, um das Programm weiter auszuführen.
   bei System.Windows.Media.Imaging.WriteableBitmap..ctor(Int32 pixelWidth, Int32 pixelHeight, Double dpiX, Double dpiY, PixelFormat pixelFormat, BitmapPalette palette)
   bei CefSharp.Wpf.Rendering.WritableBitmapRenderHandler.<>c__DisplayClass4_0.<CreateOrUpdateBitmap>b__0() in C:\projects\cefsharp\CefSharp.Wpf\Rendering\WritableBitmapRenderHandler.cs:Zeile 106.
   bei System.Windows.Threading.ExceptionWrapper.InternalRealCall(Delegate callback, Object args, Int32 numArgs)
   bei System.Windows.Threading.ExceptionWrapper.TryCatchWhen(Object source, Delegate callback, Object args, Int32 numArgs, Delegate catchHandler)

This is followed by 
> System.NullReferenceException: Der Objektverweis wurde nicht auf eine Objektinstanz festgelegt.
   bei CefSharp.Wpf.Rendering.WritableBitmapRenderHandler.<>c__DisplayClass4_0.<CreateOrUpdateBitmap>b__0() in C:\projects\cefsharp\CefSharp.Wpf\Rendering\WritableBitmapRenderHandler.cs:Zeile 119.
   bei System.Windows.Threading.ExceptionWrapper.InternalRealCall(Delegate callback, Object args, Int32 numArgs)
   bei System.Windows.Threading.ExceptionWrapper.TryCatchWhen(Object source, Delegate callback, Object args, Int32 numArgs, Delegate catchHandler)
2022-10-08 09:36:59,591 [ 1] DEBUG Betreuung.Util.DialogUtil Fehlermeldung 'Der Objektverweis wurde nicht auf eine Objektinstanz festgelegt.' öffnen ...

In [this issue](https://github.com/cefsharp/CefSharp/issues/2035#issuecomment-555296505) it's suggested to use a .net version >= 4.6. As my app is running with 4.8 this does not solve the problem in my case.

There's another [closed issue ](https://github.com/cefsharp/CefSharp/issues/3492)but it doesn't contain a "solution" (as far i can see).

What could be the reason for this exception? Is it something in CefSharp or is it a .NET issue? Is there a workaround available?

Please note that this occured on a customer's machine. Therefore detailed profiling is not possible. If needed, I can provide the cef logs.

**Comment by abid76** [comment]
Additional info: this also happens with 106.0.260.

It is not an issue with a specific pdf file. I noticed that the app's memory consumption keeps growing. E.g. it starts at 250 MB memory usage and, after two hours, it reaches 2,5 GB which then causes the exception. In this time about 200 documents (both html and pdf, all local) have been opened.

Until now it only occured when user had a 4k monitor.

If a memory snapshot is needed I have to find a way to capture it on the customer's machine.

If there is any additional data needed, please let me know.

**Reply by amaitland** [reply]
Converting to discussion for now, a new issue can be created if a memory leak is identified. 

> * **What architecture x86 or x64?**
>   x86

The additional memory requirements of a `4k` monitor using the `WPF` implementation really needs a `64bit` process.

> What could be the reason for this exception?

`CefSharp` attempted to create a `WriteableBitmap`, there was insufficient memory to do so, that's all I can say from the exception. 

> Is it something in CefSharp

Whilst it is possible, if you only see this on a `4k` display then this is unlikely. 

> is it a .NET issue?

`WritableBitmap` is used to render each frame. We have to rely on the `GC` to cleanup each `WritableBitmap` which is less than ideal when they are created in rapid succession.

- https://github.com/dotnet/wpf/issues/7177
- https://github.com/dotnet/wpf/issues/3067

> If a memory snapshot is needed I have to find a way to capture it on the customer's machine.

You'll need to profile your application to see what's actually going on.

My advice is to switch to the `x64` version. If you are unable to do so easily then at least make sure your application is `LargeAddressAware`. See https://github.com/cefsharp/CefSharp/wiki/General-Usage#win32-out-of-memory

Reuse `ChromiumWebBrowser` instances where ever possible. When you've finished with an instance make sure to call `ChromiumWebBrowser.Dispose`.

**Reply by abid76** [reply]
Thanks for helping me out! I'll look forward to switch the application to x64.

**Reply by amaitland** [reply]
Profiling your application is still probably worthwhile.
